def help_file() {
    log.info """
    #######################################################################################
    ##################### RUN QC AND SUMMARY STATS ON DOWNLOADED DATA #####################
    #######################################################################################

        Runs cutadapt on pacbio data to filter residual adapters; generates read length
        summary plot.
        
        Following Hanrahan et al. 2025 (doi.org/10.1093/g3journal/jkaf046),
        cutadapt is run with the following parameters: 
            --error-rate 0.1 
            --overlap 25 
            --match-read-wildcards 
            --revcomp 
            --discard-trimmed

        To change this, edit the ext.args line in nextflow.config

        OPTIONS:

        --yaml <PATH/TO/YAML/FILE>
                Path to the .yaml file generated by the data mapper query  
        
        --indir <PATH/TO/INPUT/DIRECTORY>
                File path to the directory where raw hifi read files are stored
                Default is './results/raw_reads/hifi'
        
        --outdir <PATH/TO/OUTPUT/DIRECTORY>
                File path to where processed hifi reads should be stored
                Default is './results/processed_reads/hifi'

        --stats <PATH/TO/OUTPUT/DIRECTORY>
                File path to where summary stats .json file should be stored
                Default is './results/qc'

        --logs <PATH/TO/OUTPUT/DIRECTORY>
            File path to where cutadapt and seqkit logs should be stored
            Default is './results/qc/pacbio_logs'

        --pacbio_adapters_fasta
                Path to .fasta file containing PacBio HiFi adapters to filter
                Default is 'assets/pacbio_adapters.fa'

    #######################################################################################
    """.stripIndent()
}


// print help file if requested
if ( params.remove('help') ) {
    help_file()
    exit 0
}

// check no unexpected parameters were specified
allowed_params = [
    // pipeline inputs
    "yaml",
    "indir",
    "outdir",
    "stats",
    "logs",
    "pacbio_adapters_fasta",

    // Pawsey options
    "max_cpus",
    "max_memory"
]

// check for unrecognised parameters
params.each { entry ->
    if ( !allowed_params.contains(entry.key) ) {
        println("The parameter <${entry.key}> is not known");
        exit 0;
    }
}

// check yaml file is specified; throw error if not
if ( !params.yaml ) { error(
    """
    ERROR: No yaml file provided: \'--yaml\'
    """
)}


include { BAM_TO_FASTQ } from './modules/bam_to_fastq.nf'
include { CUTADAPT } from './modules/cutadapt.nf'
include { READ_LENGTH_SUMMARY } from './modules/read_length_summary.nf'
include { GENERATE_STATS_FILE } from './modules/generate_stats_file.nf'


def readYAML(yamlfile) {
    return new org.yaml.snakeyaml.Yaml().load(yamlfile.text)
}

workflow {

    // read in the yaml file
    def yaml_data = readYAML(file(params.yaml))

    // pull out and format the pacbio section of the config
    pacbio_samples = yaml_data.reads.PACBIO_SMRT

    // check it exists; format data for downloading
    if ( !pacbio_samples ) { error(
        """
        No PacBio samples are listed in the config .yaml
        """
        ) 
    } else {
        pacbio_samples_reformatted = pacbio_samples
            .collectMany { pkg, pkgData ->
                pkgData.collect { file ->
                    def file_path = "${params.indir}"+file.name
                    [
                        package: pkg,
                        file_name: file.name,
                        format: file.format,
                        url: file.url,
                        md5sum: file.md5sum,
                        lane: [],
                        read: [],
                        file: file_path
                    ]

                }
            }
        pacbio_samples_ch = Channel.from(pacbio_samples_reformatted)
    }

    // check each input file exists; throw error if not
    pacbio_samples_ch
        .map { file_info ->
                if ( !file(file_info.file).exists() ) { error(
                """
                ERROR: ${file_info.file} does not exist. Check \'--indir\' is correct
                """
                )
            } 
        }

    BAM_TO_FASTQ(pacbio_samples_ch)

    // cutadapt_ch = BAM_TO_FASTQ.out.fastq

    // CUTADAPT(cutadapt_ch, "${params.pacbio_adapters_fasta}")

    // read_length_summary_ch = CUTADAPT.out.filt_fastq_gz

    // READ_LENGTH_SUMMARY(read_length_summary_ch)

    // GENERATE_STATS_FILE(CUTADAPT.out.cutadapt_log, READ_LENGTH_SUMMARY.out.summary_stats)

}