# atol-qc-raw-pacbio

Runs QC and produces summary stats on Pacbio HiFi reads
1. Convert .bam file to .fastq for downstream processing
2. Filter reads (`cutadapt`)
3. Output stats and read length distribution plot (`seqkit`)


> [!NOTE]
> Following [Hanrahan et al. 2025](https://doi.org/10.1093/g3journal/jkaf046),
> `cutadapt` is run with the following parameters by default: 
> ```
>     --error-rate 0.1
>     --overlap 25
>     --match-read-wildcards
>     --revcomp
>     --discard-trimmed
> ```
> To change this, edit the `ext.args` line in the `nexflow.config` file

## Installation

To install the latest version on Pawsey:

```
module load nextflow/25.04.6
nextflow pull amytims/atol-qc-raw-pacbio -r dev
```

## Usage

> [!IMPORTANT]
> Pipeline currently requires an atol-datamapper .yaml file to validate file inputs \
> This will be made optional in future releases

### Run QC for all PacBio reads listed in config.yaml file

```
nextflow run amytims/atol-qc-raw-pacbio \
    -r dev \
    -profile pawsey \
    --yaml <PATH/TO/DATAMAPPER/OUTPUT/YAML> \
    --indir <PATH/TO/RAW_HIFI_READS_DIRECTORY> \
    --outdir <PATH/TO/OUTPUT_HIFI_READS_DIRECTORY> \
    --stats <PATH/TO/STATS_FILES_DIRECTORY> \
    --logs <PATH/TO/LOGFILES_DIRECTORY> \
    --min_length INT
```

### Full usage

```
options:
    --help
            Show this help message and exit

    --min_length INT
            Minimum length read to output
            Default is 1, i.e. keep all reads

    --pacbio_adapters_fasta
            Path to .fasta file containing PacBio HiFi adapters to filter
            Default is 'assets/pacbio_adapters.fa'

input:
    --yaml <PATH/TO/YAML/FILE>
            Path to the .yaml file generated by the data mapper query  
    
    --indir <PATH/TO/INPUT/DIRECTORY>
            File path to the directory where raw hifi read files are stored
            Default is './results/raw_reads/hifi'

output:
    --outdir <PATH/TO/OUTPUT/DIRECTORY>
            File path to where processed hifi read files should be stored
            Default is './results/processed_reads/hifi'

    --stats <PATH/TO/OUTPUT/DIRECTORY>
            File path to where summary stats .json file(s) should be stored
            Default is './results/qc'

    --logs <PATH/TO/OUTPUT/DIRECTORY>
            File path to where cutadapt and seqkit logs should be stored
            Default is './results/qc/pacbio_logs'
```

## To-do
1. Test pipeline with added `--min_length` option, make sure it hasn't broken anything
2. Make use of .yaml file optional - just set up to print a warning message and then process all .bam files in `--indir`
3. Look into containerising pipeline for use outside of Pawsey
